### PR TITLE
psycopg2 -> psycopg3

### PR DESCRIPTION
**Pull request by rustprooflabs**
psycopg3 is in beta, I have been testing for a while now and am happy to switch over here.

The main point of note is that in psycopg3, the isolation level is now 1-indexed, per https://www.psycopg.org/psycopg3/docs/api/connections.html?highlight=isolation_level#psycopg.IsolationLevel

Screenshot from 2021-08-30 20-26-18

----

pyscopg2 was 0 indexed, per: https://kb.objectrocket.com/postgresql/setting-the-psycopg2-isolation-level-in-python-for-postgresql-transactions-648

### CODE REVIEW

**Comment by rustprooflabs** [comment]
Turns out, there's no need to mess with isolation level at all.  Have to set `autocommit=True` but that's it.

**Comment by jacopofar** [comment]
That's super interesting, I used asyncpg to query the DB populated with pgosm and it allows to transfer geometries in binary format and directly get Shapely objects while psycopg2 only uses the text protocol. Hopefully psycopg3 allows to get the same performance improvements without the need for async, I am really looking forward to it.

**Comment by rustprooflabs** [comment]
That is cool, I haven't played with the async side at all yet.  Been impressed that psycopg3 was nearly a drop-in replacement for the old.  The new connection pool is really cool too... not really applicable to this project but worth mentioning!

File: docker/db.py [hunk 1]
```diff
@@ -65,25 +70,25 @@ def pg_version_check():
 def drop_pgosm_db():
     """Drops the pgosm database if it exists."""
     sql_raw = 'DROP DATABASE IF EXISTS pgosm;'
-
     conn = get_db_conn(db_name='postgres')
-    cur = conn.cursor()
-    # Required to drop DB
-    conn.set_isolation_level(0)
```
Comment (dvarrazzo) [inline docker/db.py:72]: psycopg2 also has constant values for isolation level. but having autocommit as isolevel was just a mistake made 15 years ago :smile: https://www.psycopg.org/docs/extensions.html#isolation-level-constants

**Comment by dvarrazzo** [comment]
@jacopofar I have tried to get in touch with someone in the GIS environment but with not much success. If there is a set of standard objects to exchange geo-types between Python and Postgres I would be happy to support it in psycopg 3. If you have ideas please feel free to open a conversation on https://github.com/psycopg/psycopg/discussions. Cheers!

**Comment by jacopofar** [comment]
@dvarrazzo I've used asyncpg adapters to parse the geometries as Shapely object directly [here](https://github.com/jacopofar/geoshiny/blob/main/geoshiny/database_extract.py#L58), it automatically uses the binary protocol and when it encounters a "geometry" type data it delegates its decoding to Shapely. I found this is quite performant compared to parsing the geoJSON and the corresponding Shapely native representation is smaller than the parsed dictionary.

I can write more details in that page if you think it could be useful, I think an encoder for binary types would be amazing, to automate operations like converting a Python timestamp (I did it [here](https://github.com/jacopofar/subreddit-downloader/blob/main/src/ingest_into_postgres_psycopg3_with_copy.py#L131), but that was an old version of psycopg3 and I think is not needed), in that case it could be a default parser if Shapely is installed

**Comment by dvarrazzo** [comment]
Interesting, thank you for checking out psycopg 3 binary copy. Timestamps now are decoded well from binary format, so that function wouldn't be needed.

From what I can see, decoding those binary types is as simple as:

```python
import psycopg
from psycopg.adapt import Loader
from psycopg.pq import Format

class GeometryBinaryLoader(Loader):
    format = Format.BINARY
    def load(self, data):
        return shapely.wkb.loads(wkb)

psycopg.adapters.register_loader("box", GeometryBinaryLoader)
psycopg.adapters.register_loader("point", GeometryBinaryLoader)
...
```

That's a basic idea which should work. How to package it?

I would avoid to use shapely automatically because 1) I wouldn't add it as a dependency and 2) in case of lazy load, the failure mode in case shapely it is not installed is bad (error only at the first query that happens to use geometric types, not earlier). I would be positive to add these loaders/dumpers in the code in a module `psycopg.types.shapely` (i.e. explicitly referring to shapely, leaving open the possibility to use other geom libraries) and add a function `psycopg.types.shapely.register_shapely()` similar to [others](https://github.com/psycopg/psycopg/blob/cb0c2c117b41728e395b0bc523061a86c053bd0b/psycopg/psycopg/types/net.py#L166) dealing with objects with known oids.

Would you be interested in contributing to the feature?

**Comment by jacopofar** [comment]
Sounds like a good solution, I am looking forward to try implementing it, will do next days

**Pull request merged**